ec/internal/planmodifiers/use_state_unless_template_changed.go

```diff
@@ -20,42 +20,72 @@ package planmodifiers
 import (
 	"context"
 
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
 	"github.com/hashicorp/terraform-plugin-framework/path"
 	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
 )
 
 // Use current state for a topology's attribute if the topology's state is not nil and the template attribute has not changed
-func UseStateForUnknownUnlessTemplateChanged() planmodifier.String {
-	return useStateForUnknownUnlessTemplateChanged{}
+func UseStateForUnknownUnlessMigrationisRequired(resourceKind string) useStateForUnknownUnlessTemplateChanged {
+	return useStateForUnknownUnlessTemplateChanged{resourceKind: resourceKind}
 }
 
-type useStateForUnknownUnlessTemplateChanged struct{}
+type useStateForUnknownUnlessTemplateChanged struct {
+	resourceKind string
+}
+
+type PlanModifierResponse interface {
+	planmodifier.StringResponse | planmodifier.Int64Response
+}
 
 func (m useStateForUnknownUnlessTemplateChanged) PlanModifyString(ctx context.Context, req planmodifier.StringRequest, resp *planmodifier.StringResponse) {
-	if !resp.PlanValue.IsUnknown() {
-		return
+	useState, diags := m.UseState(ctx, req.ConfigValue, req.Plan, req.State, resp.PlanValue)
+	resp.Diagnostics.Append(diags...)
+	if useState {
+		resp.PlanValue = req.StateValue
 	}
+}
 
-	// if the config is the unknown value, use the unknown value otherwise, interpolation gets messed up
-	if req.ConfigValue.IsUnknown() {
-		return
+func (m useStateForUnknownUnlessTemplateChanged) PlanModifyInt64(ctx context.Context, req planmodifier.Int64Request, resp *planmodifier.Int64Response) {
+	useState, diags := m.UseState(ctx, req.ConfigValue, req.Plan, req.State, resp.PlanValue)
+	resp.Diagnostics.Append(diags...)
+	if useState {
+		resp.PlanValue = req.StateValue
 	}
+}
+
+func (m useStateForUnknownUnlessTemplateChanged) UseState(ctx context.Context, configValue attr.Value, plan tfsdk.Plan, state tfsdk.State, planValue attr.Value) (bool, diag.Diagnostics) {
+	var diags diag.Diagnostics
 
-	if req.StateValue.IsNull() {
```

Comment:
I think this caused the recent errors in acceptance tests.

Reply:
Hmm, yeah, that makes sense! I believe we only need this behaviour for the `instance_configuration_version`, because the remaining fields can't be `null`. Maybe I could move this condition to `PlanModifyInt64`, WDYT?

Reply:
Yes, it should work too. Though I'd prefer to explicitly control (e.g. via an argument) allowing `null` in state instead of relying on the modifier type.

Reply:
Ohh, that sounds great! That way we could specifically set `instance_configuration_version` to be a "nullable" field, but leave all of the others (including `zone_count`, which is also `Int64`) as "non-nullable". I'll add those changes to the PR 🙂